pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -230,25 +230,35 @@ private void handleMessage(Message<T> msg) {
         }
     }
 
-    private CompletableFuture<Reader<T>> readAllExistingMessages(Reader<T> reader) {
+    @Override
+    public CompletableFuture<Void> refreshAsync() {
+        return reader.thenCompose(reader -> readAllExistingMessages(reader, false));
+    }
+
+    @Override
+    public void refresh() throws PulsarClientException {
+        refreshAsync().join();
+    }
+
+    private CompletableFuture<Void> readAllExistingMessages(Reader<T> reader, boolean readTailMessages) {
         long startTime = System.nanoTime();
         AtomicLong messagesRead = new AtomicLong();
 
-        CompletableFuture<Reader<T>> future = new CompletableFuture<>();
-        readAllExistingMessages(reader, future, startTime, messagesRead);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        readAllExistingMessages(reader, future, startTime, messagesRead, readTailMessages);
         return future;
     }
 
-    private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Reader<T>> future, long startTime,
-                                         AtomicLong messagesRead) {
+    private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Void> future, long startTime,
+                                         AtomicLong messagesRead, boolean readTailMessages) {
         reader.hasMessageAvailableAsync()
                 .thenAccept(hasMessage -> {
                    if (hasMessage) {
                        reader.readNextAsync()
                                .thenAccept(msg -> {
                                   messagesRead.incrementAndGet();
                                   handleMessage(msg);
-                                  readAllExistingMessages(reader, future, startTime, messagesRead);
+                                  readAllExistingMessages(reader, future, startTime, messagesRead, readTailMessages);
```

Comment:
If I understand correctly, this PR will depend on https://github.com/apache/pulsar/pull/21270, right? Otherwise, the implementation will not conform to [what the API said](https://github.com/apache/pulsar/pull/21417/files#diff-4ec98b75d3bcad44ec9b4ee0a8853435c9f20e37880a917a544797a8942f4e41R123-R124).

Reply:
Yes, we can work on https://github.com/apache/pulsar/pull/21270 first.